helix-core/src/main/java/org/apache/helix/controller/rebalancer/AbstractRebalancer.java

```diff
@@ -417,4 +417,25 @@ public int compare(String ins1, String ins2) {
       return p1.compareTo(p2);
     }
   }
+
+  // This is for a backward compatible workaround to fix
+  // https://github.com/apache/helix/issues/940.
+  // TODO: remove the workaround once we are able to apply the simple fix without majorly
+  // TODO: impacting user's clusters.
+  protected List<String> getStablePartitionList(ResourceControllerDataProvider clusterData,
```

Comment:
Why do  we compare here again? You are saying the IdealState used here could be different from  cluster data cache?

Reply:
If the method is put in a different class, it would be safer to protect from the OO design perspective. In detail, after we put refreshStablePartitionList as private into ResourceControllerDataProvider, we don't have a mechanism to guarantee refreshStablePartitionList() is called and the refresh is done completely.

It is optional though, especially it is a workaround. If it is a big concern to you, I can remove it.